text/0000-expand-math.md

```diff
@@ -0,0 +1,103 @@
+- Feature Name: Expand `math`
+- Start Date: 2022-02-28
+- RFC PR:
+- Pony Issue:
+
+# Summary
+
+This RFC proposes expanding the standard library's `math` package and the shape it might take to leverage the full extent of Pony. There was prior discussion on [Zulip](https://ponylang.zulipchat.com/#narrow/stream/192795-contribute-to.20Pony/topic/math.20lib) years ago, but I expect some opinions to have changed.
+
+# Motivation
+
+Currently, `math` includes limited functionality. This should be expanded to include math types, constants, and present a structure with further expansion in mind. An expanded `math` library will allow a unified mathematics among Pony developers.
+
+# Detailed design
+
+The primary goals of this initial expansion are: 
+
+1. restructure the `math` library into distinct sub-packages
+2. provide common `math` data types
+
+## Structure
+
+I propose a structure of distinct subpackages including the following:
+
++ `math/big`: Arbitrary precision numbers
++ `math/series`: Mathematical series
++ `math/constant`: Mathematical constants
++ `math/rational`: `Rational` data type and related functions
++ `math/complex`: `Complex` data type and related functions
++ `math/(x,exp,etc)`: experimental additions, utilities, and effective "catch-all" for matters that do not neatly fit into other subpackages
+
+## Common Data Types
+
+As previewed above in [Structure](#structure), expanding the `math` package should include implementations of common mathematics data types. Below are some implementation proposals for those data types.
+
+### `math/big`
+
+Should include `BigInt`, `BigFloat`, and `BigDecimal` -- see [168](https://github.com/ponylang/rfcs/issues/168)
+
+### `math/series`
+
+`math/series` should include a `Series` trait which is a subclass of `Iterator`. The purpose of creating a new abstract data type is to generalize functions over mathematical series which do not make sense over iterators -- such as whether a `Series` is diverging or converging, a property that is all but meaningless for an `Iterator`.
+
+Example series include `Fibonacci` (already exists), `Pascal` (nCk), `Triangular` ({n+1}C{2}), `Square` (n^2), `Pentagonal` ({2n * (2n - 1)} / 2), etc.
+
+Current `math/Fibonacci` would go into this series package.
+
+### `math/constant`
+
+Initial values to include are those with underlying LLVM representations from the [numbers namespace](https://llvm.org/doxygen/namespacellvm_1_1numbers.html).
+
+Once these values exist in `math/constant`, they should be removed from where they are now, which is on `F32` and `F64` of [`Float`](https://github.com/ponylang/ponyc/blob/master/packages/builtin/float.pony).
```

Comment:
What's the rationale/motivation for moving them out of the `F32`/`F64` types and into a new package?

Currently the "Motivation" section of this RFC only talks about enhancing/expanding math capabilities, but this section includes taking existing functionality and moving it to a new place, so I think that warrants its own discussion of motivation. It wasn't immediately clear to me what the benefit of this is, so it probably should be stated explicitly.

Reply:
It feels weird to me that "basic datatypes" are in math rather than operations on datatypes. 

A question would also be... is it possible to move F32 etc from builtins or are they required there.

Reply:
These do not necessarily need to move, but it is what I am suggesting. For me the reason is that we have previously had new Pony users (and experiences Pony users) be surprised by the availability of these constants on `F32` and `F64`. I can expand the motivation to reflect this reasoning.

Reply:
I have changed it to say "could" not "should" as considering F32 and F64 are immutable by default (an oversight on my part) my concern for keeping them in place was misplaced. Now it is just this "surprised placement" concern, which is not a good enough reason to remove it.